File: server/schemas/shipment.py
# -*- coding: utf-8 -*-
"""판매출고 confirm REST 스키마 (Stage 5C)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class ShipLineRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    qty: float = Field(..., gt=0)
    order_detail_id: str | None = None
    item_cd: str = ""
    variety_cd: str = ""
    grade_cd: str = ""
    size_cd: str = ""
    weight: float = 0.0
    harvest_year: int = 0
    wh_cd: str = ""
    unit_price: float = Field(0.0, ge=0)


class ShipConfirmRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    ship_mode: Literal["STOCK", "DIRECT"]
    sales_dt: str = ""
    order_no: str | None = None
    custm_id: str | None = None
    rmk: str = ""
    dlvry_tp: str = ""
    ship_fee: float = Field(0.0, ge=0)
    rcv_name: str = ""
    rcv_tel: str = ""
    rcv_addr: str = ""
    dlvry_msg: str = ""
    lines: list[ShipLineRequest] = Field(min_length=1)


class ShipDetailOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    sale_detail_no: str
    order_detail_id: str | None = None
    stock_seq: int
    qty: float


class RemainingOrderOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    order_detail_id: str
    order_qty: float
    confirmed_shipped_qty: float
    remaining_order_qty: float


class ShipConfirmResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    ok: bool
    sales_no: str
    sales_status: str
    ship_mode: str
    order_no: str | None = None
    details: list[ShipDetailOut]
    order_status: str | None = None
    remaining_order_qty: float | None = None
    remaining_order: list[RemainingOrderOut] = Field(default_factory=list)

File: mobile/src/types/shipment.ts
/** Stage 6 판매출고 confirm API 계약 — stock_seq 없음 */

export type ShipMode = 'STOCK' | 'DIRECT'

export type ShipDeliveryAllocation = {
  qty: number
  rcv_name: string
  rcv_tel: string
  rcv_addr: string
  dlvry_msg?: string
  ship_fee: number
}

export type ShipConfirmLine = {
  qty: number
  order_detail_id?: string | null
  item_cd?: string
  variety_cd?: string
  grade_cd?: string
  size_cd?: string
  weight?: number
  harvest_year?: number
  wh_cd?: string
  unit_price?: number
  /** null/미전송 = legacy. 배열 = 2C STOCK 택배 */
  delivery_allocations?: ShipDeliveryAllocation[] | null
}

export type ShipConfirmRequest = {
  ship_mode: ShipMode
  sales_dt?: string
  order_no?: string | null
  custm_id?: string | null
  rmk?: string
  dlvry_tp?: string
  ship_fee?: number
  rcv_name?: string
  rcv_tel?: string
  rcv_addr?: string
  dlvry_msg?: string
  lines: ShipConfirmLine[]
}

export type ShipConfirmDetail = {
  sale_detail_no: string
  order_detail_id: string | null
  stock_seq: number
  qty: number
}

export type RemainingOrderLine = {
  order_detail_id: string
  order_qty: number
  confirmed_shipped_qty: number
  remaining_order_qty: number
}

export type ShipConfirmResponse = {
  ok: boolean
  sales_no: string
  sales_status: string
  ship_mode: ShipMode | string
  order_no: string | null
  details: ShipConfirmDetail[]
  order_status: string | null
  remaining_order_qty: number | null
  remaining_order: RemainingOrderLine[]
}